extensions/amp-video/1.0/video-iframe.js

```diff
@@ -0,0 +1,152 @@
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {Deferred} from '../../../src/utils/promise';
+import {forwardRef} from '../../../src/preact/compat';
+import {
+  useCallback,
+  useImperativeHandle,
+  useLayoutEffect,
+  useMemo,
+  useRef,
+} from '../../../src/preact';
+
+const DEFAULT_SANDBOX = [
+  'allow-scripts',
+  'allow-same-origin',
+  'allow-popups',
+  'allow-popups-to-escape-sandbox',
+  'allow-top-navigation-by-user-activation',
+].join(' ');
+
+/**
+ * Goes inside a VideoWrapper.
+ *
+ *    import {VideoIframe} from '.../video-iframe';
+ *    import {VideoWrapper} from '.../video-wrapper';
+ *    render(<VideoWrapper component={VideoIframe} ... />)
+ *
+ * Usable on the AMP layer through VideoBaseElement.
+ *
+ * @param {!VideoIframeDef.Props} props
+ * @param {{current: (T|null)}} ref
+ * @return {PreactDef.Renderable}
+ * @template T
+ */
+function VideoIframeWithRef(
+  {
+    loading = 'lazy',
+    sandbox = DEFAULT_SANDBOX,
+    muted = false,
+    controls = false,
+    origin,
+    onCanPlay,
+    onMessage,
+    makeMethodMessage,
+    ...rest
+  },
+  ref
+) {
+  const iframeRef = useRef(null);
+
+  const readyDeferred = useMemo(() => new Deferred(), []);
+
+  const postMethodMessage = useCallback(
+    (method) => {
+      if (!iframeRef.current || !iframeRef.current.contentWindow) {
+        return;
+      }
+      readyDeferred.promise.then(() => {
+        const message = makeMethodMessage(method);
+        iframeRef.current.contentWindow./*OK*/ postMessage(message, '*');
+      });
+    },
+    [readyDeferred.promise, makeMethodMessage]
+  );
+
+  useImperativeHandle(
+    ref,
+    () => ({
+      play: () => postMethodMessage('play'),
+      pause: () => postMethodMessage('pause'),
+    }),
+    [postMethodMessage]
+  );
+
+  useLayoutEffect(() => {
+    if (!onMessage) {
+      return;
+    }
+
+    /** @param {Event} event */
+    function handleMessage(event) {
+      if (
+        (origin && !origin.test(event.origin)) ||
+        event.source != iframeRef.current.contentWindow
+      ) {
+        return;
+      }
+
+      // Triggers like an HTMLMediaElement, so we give it an iframe handle
+      // to dispatch events from. They're caught from being set on {...rest} so
+      // setting onPlay, etc. props should just work.
+      onMessage({
+        // Event
+        currentTarget: iframeRef.current,
+        target: iframeRef.current,
+
+        // MessageEvent
+        data: event.data,
+      });
+    }
+
+    const {defaultView} = iframeRef.current.ownerDocument;
+
+    defaultView.addEventListener('message', handleMessage);
+
+    return () => defaultView.removeEventListener('message', handleMessage);
+  }, [onMessage, origin]);
```

Comment:
There's some danger here. The `on` callbacks are usually allocated on each render. So there's some risk w.r.t. it changing a lot. Basically, we can't expect here that the caller will play nice and always give as an invariant callback. Does it hurt us much if we run this unsubscribe/resubscribe all the time? If it does, we may need to put the last value of `onMessage` in a ref and use it inside the handler here.

Reply:
Let me know if the way this uses a ref works for the problem we mentioned.

Reply:
Yeah. I think that'd be fine. As long as the ref is updated on each render. And also we should comment it clearly that's for an unusual case.

Reply:
PTAL, I think `usePropRef` might be a pattern.